template_reports/office_renderer/loops.py

```diff
@@ -0,0 +1,235 @@
+    if not hasattr(shape, "text_frame") or not hasattr(shape.text_frame, "text"):
+        return False
+    
+    text = shape.text_frame.text
+    if not text:
+        return False
+    
+    return LOOP_END_PATTERN.search(text.strip()) is not None
+
+
+def deep_copy_slide(presentation, slide_index, slide_dict=None):
+    """
+    Create a duplicate slide and return its index in the presentation
+    
+    Since python-pptx doesn't have built-in slide cloning, we create a slide
+    with the same layout and add a reference to the original slide in 
+    slide_dict for the rendering process to use.
+    """
+    source_slide = presentation.slides[slide_index]
+    target_slide = presentation.slides.add_slide(source_slide.slide_layout)
+    
+    # If a reference dictionary is provided, store mapping
+    if slide_dict is not None:
+        slide_dict[target_slide] = source_slide
+        
+    return len(presentation.slides) - 1
+
+
+def process_loops(prs, context, perm_user, errors):
+    """
+    Process loops in the presentation:
+    - Identify loop sections (slides between %loop var in collection% and %endloop%)
+    - For each item in the collection, create a mapping between loop slides and variables
+    - Return a list of slides to process with their context info
+    """
+    # We'll use this to track which original slides each output slide corresponds to
+    slide_mapping = {}
+    
+    # First pass: identify loop sections and collect info about them
+    loop_sections = []
+    in_loop = False
+    loop_start_index = -1
+    loop_variable = None
+    loop_collection = None
+    
+    for i, slide in enumerate(prs.slides):
+        # Track whether this slide has loop directives
+        has_loop_start = False
+        has_loop_end = False
+        loop_start_var = None
+        loop_start_collection = None
+        loop_start_shapes = []
+        
+        for shape in slide.shapes:
+            # Check for loop start
+            if is_loop_start(shape):
+                loop_start_shapes.append(shape)
+                variable, collection = extract_loop_directive(shape.text_frame.text)
+                if variable and collection:
+                    has_loop_start = True
+                    loop_start_var = variable
+                    loop_start_collection = collection
+            
+            # Check for loop end
+            if is_loop_end(shape):
+                if not in_loop and not has_loop_start:
+                    errors.append(f"Error on slide {i + 1}: %endloop% without a matching loop start")
+                has_loop_end = True
+        
+        # Detect multiple loop start directives on the same slide
+        if len(loop_start_shapes) > 1:
+            errors.append(f"Error on slide {i + 1}: Multiple loop start directives on same slide")
+            continue
+        
+        # Cannot have both loop start and end on the same slide
+        if has_loop_start and has_loop_end:
```

Comment:
we CAN have both loop start end loop end on same slide. that just means we duplicate that one single slide. make sure to adjust tests if necessary!